scan-command/src/main/java/io/ballerina/scan/internal/ProjectAnalyzer.java

```diff
@@ -141,26 +140,28 @@ Map<String, List<Rule>> getExternalAnalyzers() {
         PackageResolution packageResolution = project.currentPackage().getResolution();
         ResolvedPackageDependency rootPkgNode = new ResolvedPackageDependency(project.currentPackage(),
                 PackageDependencyScope.DEFAULT);
-        List<Package> directDependencies = packageResolution.dependencyGraph().getDirectDependencies(rootPkgNode)
-                .stream().map(ResolvedPackageDependency::packageInstance).toList();
         Map<String, List<Rule>> externalAnalyzers = new HashMap<>();
-        for (Package pkgDependency : directDependencies) {
-            PackageManifest pkgManifest = pkgDependency.manifest();
-            String org = pkgManifest.org().value();
-            String name = pkgManifest.name().value();
-            String pluginName = org + FORWARD_SLASH + name;
-            if (pkgManifest.compilerPluginDescriptor().isEmpty() || !analyzerDescriptors.contains(pluginName)) {
-                continue;
-            }
-            CompilerPluginDescriptor pluginDesc = pkgManifest.compilerPluginDescriptor().get();
-            String ruleFileContent = loadRuleFileContent(pluginName, pluginDesc);
-            List<Rule> externalRules = loadExternalRules(org, name, pluginName, ruleFileContent);
-            externalAnalyzers.put(pluginDesc.plugin().getClassName(), externalRules);
-        }
+        packageResolution.dependencyGraph().getDirectDependencies(rootPkgNode).stream()
+                .map(ResolvedPackageDependency::packageInstance).forEach(pkgDependency -> {
+                    PackageManifest pkgManifest = pkgDependency.manifest();
+                    String org = pkgManifest.org().value();
+                    String name = pkgManifest.name().value();
+                    String pluginName = org + FORWARD_SLASH + name;
+                    if (pkgManifest.compilerPluginDescriptor().isEmpty()) {
+                        return;
+                    }
+                    CompilerPluginDescriptor pluginDesc = pkgManifest.compilerPluginDescriptor().get();
+                    Optional<String> ruleFileContent = loadRuleFileContent(pluginName, pluginDesc);
+                    if (ruleFileContent.isEmpty()) {
+                        return;
```

Comment:
Can we filter the dependencies using these conditions `!pkgManifest.compilerPluginDescriptor().isEmpty()` and `!ruleFileContent.isEmpty()` and then do the rest of the operations for the filtered items?

Reply:
The `loadRuleFileContent` method requires the `pluginName`, which is obtained by combining the `org` and `name` from `pkgManifest`, as well as the presence of the `compilerPluginDescriptor`. Additionally, `org` and `name` are also used in the `loadExternalRules` method. Filtering dependencies first based on these conditions would lead to code duplication, as we would need to retrieve `org` and `name` again. Therefore, will keep the current implementation to avoid redundancy.